src/haddock/libs/libgrid.py

```diff
@@ -0,0 +1,508 @@
+                output = self._find_output(line)
+                if output:
+                    self.expected_outputs.append(output)
+
+                # Process the line to adjust paths, if any
+                new_line, found_fname = self._process_line(line)
+
+                f.write(new_line)
+
+                # Collect the files that need to be in the payload
+                if found_fname:
+                    src_path = Path(found_fname)
+                    dst_path = self.loc / Path(found_fname).name
+                    shutil.copy(src_path, dst_path)
+                    self.payload_fnames.append(dst_path)
+
+    def submit(self) -> None:
+        """Interface to submit the job to DIRAC."""
+
+        try:
+            # The SUBMIT_CMD returns the job ID in stdout
+            result = subprocess.run(
+                ["dirac-wms-job-submit", f"{self.loc}/job.jdl"],
+                shell=False,
+                capture_output=True,
+                text=True,
+                cwd=self.loc,
+                check=True,
+            )
+        except subprocess.CalledProcessError as e:
+            log.error(
+                f"Job submission failed: {e}\nStdout: {e.stdout}\nStderr: {e.stderr}"
+            )
+            raise
+
+        # Add the ID to the object
+        self.id = int(result.stdout.split()[-1])
+
+        # Update the status
+        self.update_status()
+
+    def retrieve_output(self):
+        """Retrieve the output files from DIRAC."""
+        try:
+            subprocess.run(
+                ["dirac-wms-job-get-output", str(self.id)],
+                shell=False,
+                capture_output=True,
+                text=True,
+                cwd=self.loc,
+            )
+        except subprocess.CalledProcessError as e:
+            log.error(
+                f"Retrieving output failed: {e}\nStdout: {e.stdout}\nStderr: {e.stderr}"
+            )
+            raise
+
+        # NOTE: This is the output of the `job.sh`
+        self.stdout_f = Path(f"{self.loc}/{self.id}/job.out")
+        self.stderr_f = Path(f"{self.loc}/{self.id}/job.err")
+
+        # If the job failed for some reason, save the output
+        if self.stderr_f.exists():
+            dst = Path(self.wd / f"{self.id}_dirac.err")
+            shutil.copy(self.stderr_f, dst)
+
+            log.debug(f"ID stderr: {self.stderr_f.read_text()}")
+
+        # Copy the output to the expected location
+        for output_f in self.expected_outputs:
+            src = Path(f"{self.loc}/{self.id}/{output_f}")
+            dst = Path(self.wd / f"{output_f}")
+            shutil.copy(src, dst)
+
+        # The `cns.log` is the `.out` file generated by CNS
+        # Q: Is there any scenario in which this should not be compressed?
+        src = Path(f"{self.loc}/{self.id}/cns.log")
+        with open(src, "rb") as f_in, gzip.open(f"{self.cns_out_f}.gz", "wb") as f_out:
+            f_out.writelines(f_in)
+
+    def update_status(self):
```

Comment:
Missing return value
```suggestion
    def update_status(self) -> None:
```